File: erlang/src/opcode.erl
-module(opcode).

% EXPORTS

-export([opcode/1]).
-export([to_string/1]).

% INCLUDES

-include    ("include/opcode.hrl").
-include_lib("eunit/include/eunit.hrl").

% DEFINES

% API

%% @doc Translates a 5 bit value into the equivalent internal op-code
%%      representation.
opcode(16#00) -> ?RET;
opcode(16#02) -> ?JUMP;
opcode(16#03) -> ?CALL;
opcode(16#08) -> ?FETCHP;
opcode(16#0a) -> ?FETCHB;
opcode(16#0b) -> ?FETCH;
opcode(16#0c) -> ?STOREP;
opcode(16#0d) -> ?STORE_PLUS;
opcode(16#0e) -> ?STOREB;
opcode(16#0f) -> ?STORE;
opcode(16#10) -> ?MULTIPLY;
opcode(16#11) -> ?SHL;
opcode(16#12) -> ?SHR;
opcode(16#13) -> ?NOT;
opcode(16#14) -> ?PLUS;
opcode(16#15) -> ?AND;
opcode(16#16) -> ?OR;
opcode(16#17) -> ?DROP;
opcode(16#18) -> ?DUP;
opcode(16#19) -> ?POP;
opcode(16#1a) -> ?OVER;
opcode(16#1b) -> ?A;
opcode(16#1c) -> ?NOP;
opcode(16#1d) -> ?PUSH;
opcode(16#1e) -> ?BSTORE;
opcode(16#1f) -> ?ASTORE;
opcode(X)     -> ?debugFmt("UNKNOWN CODE: ~p~n",[X]), unknown.     

%% @doc Utility function to translate an op-code into a readable string.
%%
to_string(?RET)        -> "RET";
to_string(?JUMP)       -> "JUMP";
to_string(?CALL)       -> "CALL";
to_string(?FETCHP)     -> "FETCH-P";
to_string(?FETCH)      -> "FETCH";
to_string(?STOREP)     -> "STORE-P";
to_string(?STORE_PLUS) -> "STORE-PLUS";
to_string(?STOREB)     -> "STORE-B";
to_string(?STORE)      -> "STORE";
to_string(?MULTIPLY)   -> "MULTIPLY";
to_string(?SHL)        -> "SHL";
to_string(?SHR)        -> "SHR";
to_string(?NOT)        -> "NOT";
to_string(?PLUS)       -> "PLUS";
to_string(?AND)        -> "AND";
to_string(?OR)         -> "OR";
to_string(?DROP)       -> "DROP";
to_string(?DUP)        -> "DUP";
to_string(?POP)        -> "POP";
to_string(?OVER)       -> "OVER";
to_string(?A)          -> "A";
to_string(?NOP)        -> "NOP";
to_string(?PUSH)       -> "PUSH";
to_string(?BSTORE)     -> "B-STORE";
to_string(?ASTORE)     -> "A-STORE";
to_string(_)           -> "???".     

File: assembler/antlr/java/src/za/co/twyst/GA144/assembler/instructions/OpCode.java
package za.co.twyst.GA144.assembler.instructions;

public class OpCode extends Instruction {
	// CONSTANTS

	public enum OPCODE { RET       (0x00,"ret",       ";","ret"),
                         EX        (0x01,"ex",        "ex"),
	                     JUMP      (0x02,"jump",      "jump"),
	                     CALL      (0x03,"call",      "call"),
                         NEXT      (0x05,"next",      "next"),
                         IF        (0x06,"if",        "if"),
                         MINUSIF   (0x07,"-if",       "-if"),
	                     FETCHP    (0x08,"fetch-p",   "@p"),
                         FETCH_PLUS(0x09,"fetch-plus","@+"),
	                     FETCHB    (0x0a,"fetch-b",   "@b"),
                         FETCH     (0x0b,"fetch",     "@"),
                         STOREP    (0x0c,"store-p",   "!p"),
                         STORE_PLUS(0x0d,"store-plus","!+"),
	                     STOREB    (0x0e,"store-b",   "!b"),
	                     STORE     (0x0f,"store",     "!"),
                         MULTIPLY  (0x10,"multiply",  "+*","multiply"),
	                     SHL       (0x11,"shl",       "2*","shl"),
	                     SHR       (0x12,"shr",       "2/","shr"),
	                     NOT       (0x13,"not",       "-"),
	                     PLUS      (0x14,"plus",      "+"),
	                     AND       (0x15,"and",       "and"),
	                     OR        (0x16,"xor",       "or","xor"),
	                     DROP      (0x17,"drop",      "drop"),
	                     DUP       (0x18,"dup",       "dup"),
	                     POP       (0x19,"pop",       "pop"),
	                     OVER      (0x1a,"over",      "over"),
	                     A         (0x1b,"a",         "a"),
	                     NOP       (0x1c,"nop",       ".","nop"),
	                     PUSH      (0x1d,"push",      "push"),
	                     BSTORE    (0x1e,"b-store",   "b!"),
	                     ASTORE    (0x1f,"a-store",   "a!");
	                    
	                     public final int      code;
	                     public final String[] mnemonic;
	                     public final String   string;
	                     public final boolean  slot3;
	                     
	                     private OPCODE(int code,String string,String...mnemonic) {
	                    	 this.code     = code;
	                    	 this.slot3    = (code & 0x03) == 0x00;
	                    	 this.mnemonic = mnemonic;
	                    	 this.string   = string;
	                     }
	}
	
	// INSTANCE VARIABLES
	
	public final OPCODE opcode;
	
	// CONSTRUCTOR
	
	public OpCode(OPCODE opcode) {
		// ... validate 
		
		if (opcode == null) {
			throw new IllegalArgumentException("Invalid OPCODE");
		}
		
		// ... initialise
		
		this.opcode = opcode;
	}
	                     
	// *** Object ***
	
	@Override
	public String toString() {
		return opcode.string;
	}

}
